src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java

```diff
@@ -1107,33 +1126,73 @@ public Image call() {
-            if (getLargeIcon) {
-                largeIcon = icon;
-            } else {
-                smallIcon = icon;
-            }
         }
         return icon;
     }
 
+    /**
+     * @return The icon image of specified size used to display this shell folder
+     */
+    public Image getIcon(int size) {
+        return invoke(() -> {
+            Image newIcon = null;
+            if (isLink()) {
+                Win32ShellFolder2 folder = getLinkLocation(false);
+                if (folder != null && folder.isLibrary()) {
+                    return folder.getIcon(size);
+                }
+            }
+            Map<Integer, Image> multiResolutionIcon = new HashMap<>();
+            int start = size > MAX_QUALITY_ICON ? ICON_RESOLUTIONS.length - 1 : 0;
+            int increment = size > MAX_QUALITY_ICON ? -1 : 1;
+            int end = size > MAX_QUALITY_ICON ? -1 : ICON_RESOLUTIONS.length;
+            for (int i = start; i != end; i+=increment) {
+                int s = ICON_RESOLUTIONS[i];
+                if (size < MIN_QUALITY_ICON || size > MAX_QUALITY_ICON
+                        || (s >= size/2 && s <= size*2)) {
+                    long hIcon = extractIcon(getParentIShellFolder(),
+                            getRelativePIDL(), s, false);
+
+                    // E_PENDING: loading can take time so get the default
+                    if (hIcon <= 0) {
+                        hIcon = extractIcon(getParentIShellFolder(),
+                                getRelativePIDL(), s, true);
+                        if (hIcon <= 0) {
+                            if (isDirectory()) {
+                                return getShell32Icon(FOLDER_ICON_ID, size);
+                            } else {
+                                return getShell32Icon(FILE_ICON_ID, size);
+                            }
+                        }
+                    }
+                    newIcon = makeIcon(hIcon, s);
+                    disposeIcon(hIcon);
+
+                    multiResolutionIcon.put(s, newIcon);
+                    if (size < MIN_QUALITY_ICON || size > MAX_QUALITY_ICON) {
+                        break;
+                    }
+                }
+            }
+            return new MultiResolutionIconImage(size, multiResolutionIcon);
+        });
+    }
+
     /**
      * Gets an icon from the Windows system icon list as an {@code Image}
      */
     static Image getSystemIcon(SystemIcon iconType) {
         long hIcon = getSystemIcon(iconType.getIconID());
-        Image icon = makeIcon(hIcon, true);
+        Image icon = makeIcon(hIcon, LARGE_ICON_SIZE);
         disposeIcon(hIcon);
         return icon;
     }
 
     /**
      * Gets an icon from the Windows system icon list as an {@code Image}
      */
-    static Image getShell32Icon(int iconID, boolean getLargeIcon) {
+    static Image getShell32Icon(int iconID, int size) {
         boolean useVGAColors = true; // Will be ignored on XP and later
```

Comment:
I cannot see where `useVGAColors` is used. Since the supported OS are later than XP, it can be removed, can't it?

Reply:
I think so. Fixed.